tofhir-engine/src/main/scala/io/tofhir/engine/data/write/FileSystemWriter.scala

```diff
@@ -18,15 +21,129 @@ class FileSystemWriter(sinkSettings: FileSystemSinkSettings) extends BaseFhirWri
     import spark.implicits._
     logger.debug("Created FHIR resources will be written to the given URL:{}", sinkSettings.path)
     sinkSettings.sinkType match {
+      // Handle cases where the file format is either NDJSON, PARQUET, or DELTA_LAKE,
+      // and the output needs to be partitioned by FHIR resource type.
+      case SinkFileFormats.NDJSON | SinkFileFormats.PARQUET | SinkFileFormats.DELTA_LAKE if sinkSettings.partitionByResourceType =>
+        // Transform the DataFrame (df) to include the resource type.
+        // parsedDF will have the following structure:
+        // +------------+-------------------------+
+        // | resourceType | mappedResourceJson    |
+        // +------------+-------------------------+
+        // | Patient      | {"resourceType": "... |
+        // | Patient      | {"resourceType": "... |
+        // +------------+-------------------------+
+        val parsedDF = df
+          .map(row => {
+            // extract `mappedResource` and parse the resourceType from the JSON.
+            val mappedResourceJson = row.mappedResource.get
```

Comment:
We parse the JSON string in the dataframe to find the resource type at this point. What if we put the resource type into FhirMappingResult and set it before we serialize the JObject to String within MappingTaskExecutor? If we do that, I believe we can remove this parsing and improve performance.

What do you think?

Reply:
I'll extend the `FhirMappingResult` model by adding an optional `resourceType` field. It will significantly improve the performance, as we can use this field directly to group the DataFrame.

Reply:
Done https://github.com/srdc/tofhir/commit/63cb8a99d2a1cf33ca76358739c099f491428907